lib/Sema/CSDiagnostics.h

```diff
@@ -759,6 +759,23 @@ class ContextualFailure : public FailureDiagnostic {
   void tryComputedPropertyFixIts(Expr *expr) const;
 };
 
+/// Diagnose mismatches relating to tuple destructuring.
+class TupleContextualFailure final : public ContextualFailure {
+public:
+  TupleContextualFailure(Expr *root, ConstraintSystem &cs, Type lhs, Type rhs,
+                         ConstraintLocator *locator)
+      : ContextualFailure(root, cs, lhs, rhs, locator) {}
+
+  bool diagnoseAsError() override;
+
+  bool isNumElementsMismatch() const {
+    auto lhsTy = dyn_cast<TupleType>(getFromType().getPointer());
+    auto rhsTy = dyn_cast<TupleType>(getToType().getPointer());
```

Comment:
How about `get{From, To}Type()->getAs<TupleType>()`?

Reply:
Since you assert they’re non zero below, even better to use ->castTo()